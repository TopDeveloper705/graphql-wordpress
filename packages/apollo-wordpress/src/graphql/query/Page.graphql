#import "../fragment/ContentNode_content.graphql"
#import "../fragment/Media_media.graphql"

query Page($slug: String!) {
  viewer {
    page(slug: $slug) {
      id
      slug
      title {
        raw
      }
      content {
        data {
          ...ContentNode_content
        }
      }
      featuredMedia {
        ... on Image {
          sourceUrl
        }
        ...Media_media
      }
    }
  }
}
