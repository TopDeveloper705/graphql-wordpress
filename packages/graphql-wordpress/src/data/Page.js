import { toGlobalId } from 'graphql-relay';

class Page {
  getID() {
    return toGlobalId(this.constructor.name, this.id);
  }

  static getEndpoint() {
    return process.env.WP_PAGES_ENDPOINT || 'graphql/v1/pages';
  }
}

export default Page;
