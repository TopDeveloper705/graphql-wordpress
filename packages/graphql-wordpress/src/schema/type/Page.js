import { GraphQLObjectType, GraphQLInt } from 'graphql';
import PostInterface from 'interface/Post';
import { globalIdField, slug, guid, link } from 'field/identifier';
import { title, content, excerpt } from 'field/content';
import { date, modified } from 'field/date';
import metaField from 'field/meta';
import { type, template } from 'field/post';
import { commentStatus, pingStatus } from 'field/status';
import { featuredMedia } from 'field/media';
import author from 'field/author';
import { registerNodeType, NodeInterface } from 'type/relayNode';

const PageType = new GraphQLObjectType({
  name: 'Page',
  description: 'An object.',
  interfaces: [PostInterface, NodeInterface],
  isTypeOf(page) {
    return page.type === 'page';
  },
  fields: () => ({
    id: globalIdField(),
    ...date,
    ...guid,
    ...modified,
    ...slug,
    ...type,
    ...link,
    ...title,
    ...content,
    ...excerpt,
    ...commentStatus,
    ...pingStatus,
    ...template,
    ...author,
    featuredMedia: featuredMedia(),
    meta: metaField(),
    // extra page fields
    parent: {
      type: PageType,
      description: 'The ID for the parent of the object.',
      resolve: (page, args, context, { rootValue: { loaders: { Page } } }) =>
        page.parent > 0 ? Page.load(page.parent) : null,
    },
    menuOrder: {
      type: GraphQLInt,
      description: 'The order of the object in relation to other object of its type.',
    },
  }),
});

registerNodeType(PageType);

export default PageType;
