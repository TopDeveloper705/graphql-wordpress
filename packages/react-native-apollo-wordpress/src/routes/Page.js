import React, { Component } from 'react';
import { graphql } from 'react-apollo';
import { StyleSheet, Text, View, Image } from 'react-native';
import PageQuery from 'graphql/Page_Query.graphql';
import Error from 'components/Error';
import Loading from 'components/Loading';

/* eslint-disable react/prop-types */

const styles = StyleSheet.create({
  container: {
    flex: 2,
    backgroundColor: '#fff',
    paddingLeft: 10,
    paddingRight: 10,
  },

  title: {
    fontSize: 24,
    fontWeight: 'bold',
    paddingTop: 10,
    paddingBottom: 10,
  },

  imageWrap: {
    flex: 1,
    flexDirection: 'row',
  },

  image: {
    flex: 1,
    width: null,
    height: null,
  },
});

@graphql(PageQuery, {
  options: ({ match: { params: { slug } } }) => ({
    variables: {
      slug,
    },
  }),
})
export default class Page extends Component {
  render() {
    const { data: { loading, error } } = this.props;
    if (error) {
      return <Error />;
    } else if (loading) {
      return <Loading />;
    }

    const { viewer: { page } } = this.props.data;

    return (
      <View style={styles.container}>
        <Text style={styles.title}>{page.title.raw}</Text>
        {page.featuredMedia &&
          page.featuredMedia.source_url && (
            <View style={styles.imageWrap}>
              <Image
                style={styles.image}
                source={{ uri: page.featuredMedia.source_url }}
                resizeMode="contain"
              />
            </View>
          )}
      </View>
    );
  }
}
