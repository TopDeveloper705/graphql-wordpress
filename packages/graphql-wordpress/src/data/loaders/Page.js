import Dataloader from 'dataloader';
import fetchData from 'data/utils';
import Page from 'data/Page';

// Dataloader expects IDs that can be read by the REST API

export default function getPageLoaders() {
  const endpoint = Page.getEndpoint();
  const pageLoader = new Dataloader(ids =>
    fetchData(endpoint, {
      qs: { include: ids, orderby: 'include', per_page: 100 },
    }).then(({ data: { body } }) => body)
  );
  const slugLoader = new Dataloader(slugs =>
    fetchData(endpoint, { qs: { slug: slugs } })
      .then(({ data: { body } }) => body)
      // the REST API does not order by FIELD(slug, ....) yet
      .then(pages => slugs.map(slug => pages.find(page => slug === page.slug)))
  );
  return {
    load: async id => {
      const data = await pageLoader.load(id);
      return data ? Object.assign(new Page(), data) : null;
    },
    loadBySlug: async slug => {
      const data = await slugLoader.load(slug);
      return data ? Object.assign(new Page(), data) : null;
    },
  };
}
