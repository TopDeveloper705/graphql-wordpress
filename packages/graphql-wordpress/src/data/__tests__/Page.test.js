import Page from 'data/Page';

const backupEnv = Object.assign({}, process.env);

afterAll(() => {
  process.env = backupEnv;
});

describe('Test Page data access', () => {
  test('Get endpoint', () => {
    expect(Page.getEndpoint()).toMatchSnapshot();
  });

  test('Get backup endpoint', () => {
    delete process.env.WP_PAGES_ENDPOINT;
    expect(Page.getEndpoint()).toMatchSnapshot();
  });
});
