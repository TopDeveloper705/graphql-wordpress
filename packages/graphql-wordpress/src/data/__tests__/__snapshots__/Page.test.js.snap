// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Page data access Get backup endpoint 1`] = `"graphql/v1/pages"`;

exports[`Test Page data access Get endpoint 1`] = `"graphql/v1/pages"`;
